CHERI-RISC-V TLB fill fix

nwf opened the pull request:
We need to assert `PAGE_SC_TRAP` if either `PTE_CW` or `PTE_CD` are clear, as otherise we can end up caching a TLB entry without `SC_TRAP` for a PTE that's merely got `PTE_CW` without `PTE_CD` turned on.

This isn't the greatest fix in the world, but it does have the possible defense of mimicking the existing logic for `PTE_D`.  We'd rather not take the trap if all we need to do is set CD, but I think we lack hooks in the right place, and we'd need to do the AMO thing anyway.

jrtc27 committed:
Split out new cc128m Morello format from cc128 and refactor tests

This allows both variants to coexist in the same namespace, rather than
overloading the meaning of cc128.

Note that previously ^_morello$ was used to match the Morello test
binaries, which never matched, and so all the Morello test binaries were
in fact just copies of the normal CHERI ones, testing normal CHERI-128
(and a bit of CHERI-64) instead. Morello variants are now properly
built, and CHERI-64 is more thoroughly tested, though Morello still has
a large number of disabled tests compared with normal CHERI.

nwf-msr committed:
CHERI-RISC-V: PTE_CW/CD bugfix

We need to assert PAGE_SC_TRAP if either PTE_CW or PTE_CD are clear, as
otherise we can end up caching a TLB entry without SC_TRAP for a PTE
that's merely got PTE_CW without PTE_CD turned on.

This is similar to the existing logic around PTE_D, though this commit adds some
commentary clarifying the similarity.  I assume the PTE_W/PTE_D test is phrased
the way it is for optimization.

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -716,11 +716,13 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+             * PTE_W or PTE_D are clear, force a TLB miss on later writes to
+             * update the dirty bit
+             */
+            if ((pte & PTE_W) && (pte & PTE_D)) {
```
This diverges from upstream, where it's:
```
            if ((pte & PTE_W) &&
                    (access_type == MMU_DATA_STORE || (pte & PTE_D))) {
```
though I don't think it's ever possible to get here with `access_type == MMU_DATA_(CAP_)STORE` but not `pte & PTE_D`, as either you hit the "atomically set PTE_D" case or you hit one of your RISCV_PTE_TRAPPY cases and bail out early. I therefore think this is an entirely no-op change and so we should stick to extending upstream's approach.

nwf replied:
Yes, this is a no-op change because `access_type == MMU_DATA_STORE` implies `pte & PTE_D`.  I can revert to upstream if that'd make you happier.

jrtc27 commented on target/riscv/cpu_helper.c:
```diff
@@ -743,7 +745,18 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
+             * PAGE_SC_TRAP is off the TCG-native path and so will actaully
+             * trigger a trap when all we really want is a new lookup, but
+             * it beats failing to update CD if the first access to a page
+             * isn't a cap store.
```
I mean, it's an improvement, because we have software emulation of CD for our various soft cores, but it's not great, and it's supposed to be that if hardware dirty tracking is supported then so is capability dirty tracking (though at a glance it doesn't look like we actually specify that). Can we not just fix this properly so PAGE_SC_TRAP leads to a TLB miss like a missing PAGE_WRITE (or invert it so a missing PAGE_CAP_WRITE leads to a TLB miss)? The LC handling is awkward due to the additional state, but SC doesn't have to be so messy.

nwf replied:
After 818c573e88d64b042225ceb2515e04888b7b8eff, this does the right thing.  I've updated the comment to match.

LawrenceEsswood reviewed (approved):
LGTM

LawrenceEsswood commented:
Slso w.r.t. to your comment. The guest should never take a trap if all that needs to happen is CD being set. The page tables will have to be walked again, but then again that's always going to be required unless have have our soft-tlb entries point to the PTEs they came from.

jrtc27 committed:
git subrepo push target/cheri-common/cheri-compressed-cap

subrepo:
  subdir:   "target/cheri-common/cheri-compressed-cap"
  merged:   "6762a19ba9"
upstream:
  origin:   "https://github.com/CTSRD-CHERI/cheri-compressed-cap"
  branch:   "master"
  commit:   "6762a19ba9"
git-subrepo:
  version:  "0.4.1"
  origin:   "???"
  commit:   "???"

arichardson committed:
[Morello] Don't assert jump targets are representable (#195)

Morello checks the bounds at the jump target, so it is possible that the
target PCC is unrepresentable after a jump. I believe this should fix
a "Target addr must be representable" assertion that I saw a while ago.

jrtc27 committed:
Clean up TCG instruction logging printf handling a little

Currently we conflate qemu_log_printf_valid_entries and printf_used_ptr
handling, but the former is a dynamic TCG variable that is relied upon
being cleared after printing and the latter is a bit of tracking used
during translation. Instead, clear the former as we dump in the helper
rather than require generated code at every site to clear it. This also
allows it to no longer be exposed outside the logging implementation.

We can also do the same with printf_used_ptr and keep its handling
hidden within the logging implementation rather than split between there
and translation.

jrtc27 committed:
Don't let PCC fault squash log of previous instruction

The current contract is that the next TB prints the final instruction
(which may have faulted) of the previous TB. However, the current
implementation of this puts the first flush after the PCC checks, and
this is particularly apparent when stuck in a PCC fault loop as the
logging stops after the penultimate instruction, never printing anything
more. Hoist the first case to the very beginning of the TB to fix this.

jrtc27 committed:
morello: Fix missing register change log for C64 ADR/ADRP/ADRDP

The gen_cap_*_fast just do the operation, they don't mark the register
as modified, so we need to do that ourselves as is done elsewhere for
similar cases.

jrtc27 committed:
morello: Fix missing register change log for C64 BL

nwf commented:
Bah, wrong base branch?
